### PR TITLE
MAILET-114 Refactor ToProcessor

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java [hunk 1]
```diff
@@ -42,62 +44,46 @@
...
+
+    @Override
+    public String getMailetInfo() {
+        return "ToProcessor Mailet";
+    }
+
+    @Override
     public void service(Mail mail) throws MessagingException {
-        if (isDebug) {
+        if (debug) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: we should create a ticket to remove these debug flags, loggers are good at that

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java [hunk 2]
```diff
@@ -42,62 +44,46 @@
...
-    /**
-     * Deliver a mail to the processor.
-     *
-     * @param mail the mail to process
-     *
-     * @throws MessagingException in all cases
-     */
+    private boolean isDebug() {
+        String parameter = getInitParameter("debug");
+        return (parameter == null) ? false : Boolean.valueOf(parameter);
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: Usually we don't use the ternary operator.

Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: You can use directly : 

```
return getInitParameter("debug", false);
```
I guess...

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java [hunk 3]
```diff
@@ -42,62 +44,46 @@
...
+
+    @Override
     public void service(Mail mail) throws MessagingException {
-        if (isDebug) {
+        if (debug) {
             log(String.format("Sending mail %s to %s", mail, processor));
         }
         mail.setState(processor);
-        if (noticeText != null) {
+        setErrorMessage(mail);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: Naming here is confusing... The way I understand it is it set an error message on every mail. I need to read the actual method to understand that it is only setted if the notice is present.

Maybe wrapping call in if block will make it easier to read.
Maybe naming is enough (setOptionalErrorMessage)

By the way, is it a notice or an error message ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java [hunk 4]
```diff
@@ -42,62 +44,46 @@
...
+    }
+
+    private void setErrorMessage(Mail mail) {
+        if (noticeText.isPresent()) {
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]:
```
mail.setErrorMessage(Optional.ofNullable(mail.getErrorMessage())
    .transform(new Function<String, String>() {
        String apply(String input) {
            return String.formatString.format("%s\r\n%s", mail.getErrorMessage(), noticeText.get());
        }).orElse(noticeText.get());
```
To do it in more guava way ?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java [hunk 5]
```diff
@@ -20,92 +20,118 @@
...
-            mci.setProperty("notice", notice);
-        }
-        mailet.init(mci);
+    @Test
+    public void serviceShouldSetTheStateOfTheMail() throws MessagingException {
+        String processor = "error";
+        mailetConfig.setProperty("processor", processor);
+        mailet.init(mailetConfig);
+
+        Mail mail = new FakeMail();
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: With MAILET-111 I wrote it becames : 

```
Mail mail = FakeMail.builder().recipients(...).build;
```

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java [hunk 6]
```diff
@@ -20,92 +20,118 @@
...
     }
 
-    // test if ToProcessor works
     @Test
-    public void testValidToProcessor() throws MessagingException {
-        setProcessor("error");
-        setNotice("error in message");
-        setupMockedMail(mockedMimeMessage);
-        setupMailet();
+    public void serviceShouldSetTheErrorMessageOfTheMailWhenNull() throws MessagingException {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: When Null ?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java [hunk 7]
```diff
@@ -20,92 +20,118 @@
...
+        mailet.service(mail);
 
-        mailet.service(mockedMail);
+        assertThat(mail.getErrorMessage()).isEqualTo(notice);
+    }
 
-        Assert.assertEquals(processor, mockedMail.getState());
-        Assert.assertEquals(notice, mockedMail.getErrorMessage());
+    @Test
+    public void serviceShouldAppendTheErrorMessageOfTheMailWhenSome() throws MessagingException {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: WhenSomeErrorMessageOnMail

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java [hunk 8]
```diff
@@ -42,62 +44,46 @@
...
             if (mail.getErrorMessage() == null) {
-                mail.setErrorMessage(noticeText);
+                mail.setErrorMessage(noticeText.get());
             } else {
-                mail.setErrorMessage(String.format("%s\r\n%s", mail.getErrorMessage(), noticeText));
+                mail.setErrorMessage(String.format("%s\r\n%s", mail.getErrorMessage(), noticeText.get()));
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: I don't think it is our responsability to compose error messages on mails...
It is so easy to have a caller loosing previous error messages...
However maybe it deserve it's own ticket...

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java (hunk 1 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: https://issues.apache.org/jira/browse/MAILET-126

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java (hunk 2 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: Right

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java (hunk 3 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java (hunk 4 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: No more need

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java (hunk 5 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: Ok, i will rebase my PR.
The bad point is that there's still a public constructor on FakeMail; and it's also not immutable.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java (hunk 6 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: OK

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java (hunk 7 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ToProcessorTest.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java (hunk 8 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ToProcessor.java]: We will have to think about the attribute usage later, we have to learn more about mailets behaviour.
I think we will encounter other questions like this one, we may handle them in a second step.

**Comment by aduprat** [comment]
Rebased

**Comment by mbaechler** [comment]
👍

**Comment by rouazana** [comment]
:+1:

**Pull request merged**